### ISSUE TITLE
Mypy doesn't understand func `type()` means

### ISSUE THREAD

**Issue by WangGithubUser** [python/mypy#16171]
**Bug Report**

Mypy dosen't understand what `type()` function means.

**To Reproduce**

At https://mypy-play.net/?mypy=master&python=3.11&gist=6fb5472d9376a3a1c0a25682911621d5.

**Expected Behavior**

This should pass without false positive `error: Unsupported operand types for + ("int" and "str")  [operator]`

**Actual Behavior**
Output
```
main.py:11: error: Unsupported operand types for + ("int" and "str")  [operator]
main.py:11: note: Left operand is of type "int | str"
main.py:13: error: Unsupported operand types for + ("str" and "int")  [operator]
main.py:13: note: Left operand is of type "int | str"
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

(In mypy playground)
- Mypy version used:master branch
- Mypy command-line flags:None
- Mypy configuration options from `mypy.ini` (and other config files):None
- Python version used:Python 3.11

**Comment by hauntsaninja** [comment]
mypy and other Python static type checkers don't reason about narrowing across functions like this. Use a PEP 647 TypeGuard for this kind of thing (see also PEP 724)
